src/main/java/com/atlassian/bitbucket/jenkins/internal/trigger/register/BitbucketWebhookHandler.java

```diff
@@ -76,93 +92,166 @@ private void deleteWebhooks(List<BitbucketWebhook> webhooks) {
+                .filter(hook -> hook.getEvents().containsAll(toSubscribe.stream()
+                        .map(BitbucketWebhookEvent::getEventId).collect(Collectors.toSet())))
                 .peek(hook -> LOGGER.info("Found an existing webhook - " + hook))
                 .findFirst();
     }
 
     /**
      * Returns the correct webhook event to subscribe to.
-     *
      * For Mirror sync event, the input request should point to mirror.
+     * 1. If the webhook capability does not contain mirror sync, this version with webhooks at least supports ref change
      * For Repo ref event,
      * 1. Input request does not point to mirrors
-     * 2. If webhook capability does not contains mirror sync, we still subscribe to repo ref.
-     *
-     * If the webhook capability is not there, we still subscribe to repo ref event.
+     * 2. Since this version supports webhooks, it will definitely support pull requests and ref changes
+     * 3. We subscribe events depending on if we want to only listen for ref changes, only pull request events or both
+     * If we subscribe to pull request events, we subscribe to all of these: when pull requests are opened, closed, deleted or merged.
      *
      * @param request the input request
      * @return the correct webhook event
      */
-    private BitbucketWebhookEvent getEvent(WebhookRegisterRequest request) {
+    private Collection<BitbucketWebhookEvent> getEvents(WebhookRegisterRequest request) {
+        Collection<BitbucketWebhookEvent> supportedEvents = new HashSet<>();
         if (request.isMirror()) {
             try {
                 BitbucketWebhookSupportedEvents events = serverCapabilities.getWebhookSupportedEvents();
                 Set<String> hooks = events.getApplicationWebHooks();
-                if (hooks.contains(MIRROR_SYNCHRONIZED_EVENT.getEventId())) {
-                    return MIRROR_SYNCHRONIZED_EVENT;
+                if (hooks.contains(MIRROR_SYNCHRONIZED.getEventId())) {
+                    supportedEvents.add(MIRROR_SYNCHRONIZED);
                 } else if (hooks.contains(REPO_REF_CHANGE.getEventId())) {
-                    return REPO_REF_CHANGE;
-                } else {
-                    throw new WebhookNotSupportedException("Remote server does not support the required events.");
+                    supportedEvents.add(REPO_REF_CHANGE);
                 }
-            } catch (BitbucketMissingCapabilityException exception) {
-                return REPO_REF_CHANGE;
+            } catch (BitbucketMissingCapabilityException exception) { //version doesn't support webhooks but support ref change & pr
+                    supportedEvents.add(REPO_REF_CHANGE);
+            }
+        } else {
+            if (request.isTriggerOnRefChange()) {
+                supportedEvents.add(REPO_REF_CHANGE);
+            }
+            if (request.isTriggerOnPullRequest()) {
+                supportedEvents.add(PULL_REQUEST_DECLINED);
```

Comment:
/nit 
```suggestion
                supportedEvents.addAll(ALL_PULL_REQUEST_EVENTS);
```
Or do away with the list entirely and explicitly list supported events in the `process` method (I don't mind either way, but we should be consistent)